projects/RabbitMQ.Client/client/impl/WireFormatting.cs

```diff
@@ -700,39 +718,74 @@ public static int WriteShort(Span<byte> span, ushort val)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe int WriteShortstr(Span<byte> span, string val)
+        public static int WriteShortstr(Span<byte> span, string val)
         {
-            int maxLength = span.Length - 1;
-            if (maxLength > byte.MaxValue)
-            {
-                maxLength = byte.MaxValue;
-            }
-            fixed (char* chars = val)
-            fixed (byte* bytes = &span.Slice(1).GetPinnableReference())
+            int bytesWritten = 0;
+            if (!string.IsNullOrEmpty(val))
             {
+                int maxLength = span.Length - 1;
+                if (maxLength > byte.MaxValue)
+                {
+                    maxLength = byte.MaxValue;
+                }
+#if NETCOREAPP
                 try
                 {
-                    int bytesWritten = Encoding.UTF8.GetBytes(chars, val.Length, bytes, maxLength);
-                    span[0] = (byte)bytesWritten;
-                    return bytesWritten + 1;
+                    bytesWritten = UTF8.GetBytes(val, span.Slice(1, maxLength));
                 }
                 catch (ArgumentException)
                 {
                     return ThrowArgumentOutOfRangeException(val, maxLength);
                 }
+#else
+                unsafe
+                {
+                    fixed (char* chars = val)
+                    {
+                        try
+                        {
+                            fixed (byte* bytes = span.Slice(1))
+                            {
+                                bytesWritten = UTF8.GetBytes(chars, val.Length, bytes, maxLength);
+                            }
+                        }
+                        catch (ArgumentException)
+                        {
+                            return ThrowArgumentOutOfRangeException(val, maxLength);
```

Comment:
Would we loose perf if we preserved the original exception as an inner for more transparency?

Reply:
Don't think so. We'd have to replace the maxLength parameter since there's a missing overload for ArgumentOutOfRangeException to accept both an inner exception as well as setting the value.